### PR TITLE
Feature/addressees2

### PR DESCRIPTION
**Pull request by tatianaIljina**

### CODE REVIEW

File: Services/Utilities/Addressees/Addressees.API/Controllers/DistributionGroupController.cs [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+        [ProducesResponseType(StatusCodes.Status201Created)]
+        [ProducesResponseType(StatusCodes.Status204NoContent)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+        [HttpPost]
+        public ActionResult Post([FromBody] DistributionGroupViewModel model)
+        {
+            if (model == null)
+            {
+                return this.NoContent();
```
Comment (Anny777) [inline Services/Utilities/Addressees/Addressees.API/Controllers/DistributionGroupController.cs:67]: return bad request

File: Services/Utilities/Addressees/Addressees.API/Controllers/DistributionGroupController.cs [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+            var region = this.regionRepository.Find(model.DistributionRegion);
+            if (region == null)
+            {
+                return this.NotFound($"The region {model.DistributionRegion} wasn't found");
+            }
+
+            var res = this.repository.Create(model);
+            if (res != null)
+            {
+                return this.Created(new Uri($"/api/DistributionGroup?id={res.Id}", UriKind.Relative), new { res });
```
Comment (Anny777) [inline Services/Utilities/Addressees/Addressees.API/Controllers/DistributionGroupController.cs:79]: Retutrn created at action: 
Example:
var idCreatedObj = new { id = IdObj};
return this.CreatedAtAction(nameof(this.Get), idCreatedFile, idCreatedFile);

File: Services/Utilities/Addressees/Addressees.API/Controllers/DistributionGroupController.cs [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+            }
+
+            return this.StatusCode(500, $"�������������� ������. �� ������� �������� ��������.");
```
Comment (Anny777) [inline Services/Utilities/Addressees/Addressees.API/Controllers/DistributionGroupController.cs:82]: Check the encoding file. Should be UTF-8

File: Services/Utilities/Addressees/Addressees.API/Controllers/DistributionGroupController.cs [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+        /// <response code="200">Returns distribution groups.</response>
+        /// <response code="204">If filterOptions is empty.</response>
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status204NoContent)]
+        [HttpGet("GetGroups")]
+        public ActionResult<ListResponse<DistributionGroupViewModel>> GetGroups([FromQuery]FilterOptions filterOptions)
+        {
+            if (filterOptions == null)
+            {
+                return this.NoContent();
```
Comment (Anny777) [inline Services/Utilities/Addressees/Addressees.API/Controllers/DistributionGroupController.cs:99]: return bad request

File: Services/Utilities/Addressees/Addressees.API/Controllers/DistributionGroupController.cs [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+            }
+
+            var data = this.repository.GetList(filterOptions.Skip(), filterOptions.Take(), filterOptions.QuickSearch, true, filterOptions.Filter).ToList();
+            var count = this.repository.Count(0, 0, filterOptions.QuickSearch, true, filterOptions.Filter);
+            var res = data.Select(r => this.ConvertToViewModel(r)).ToArray();
```
Comment (Anny777) [inline Services/Utilities/Addressees/Addressees.API/Controllers/DistributionGroupController.cs:104]: what happens if this.ConvertToViewModel(r) returned null?

File: Services/Utilities/Addressees/Addressees.API/Controllers/DistributionRegionController.cs [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+        /// <param name="id">The identifier.</param>
+        /// <returns>The result.</returns>
+        /// <response code="200">Deleted successfully.</response>
+        /// <response code="404">Distribution region wasn't found.</response>
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        [HttpDelete("{id}")]
+        public ActionResult Delete(Guid id)
+        {
+            var entity = this.regionRepository.Find(id.ToString());
```
Comment (Anny777) [inline Services/Utilities/Addressees/Addressees.API/Controllers/DistributionRegionController.cs:73]: Why Find have input parameter string? Why not guid?

File: Services/Utilities/Addressees/Addressees.API/Controllers/DistributionRegionController.cs [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+        /// <response code="500">Unexpected error.</response>
+        [ProducesResponseType(StatusCodes.Status201Created)]
+        [ProducesResponseType(StatusCodes.Status204NoContent)]
+        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+        [HttpPut]
+        public ActionResult<string> Put([FromBody] Addressees.API.Models.DistributionRegion value)
+        {
+            if (value == null)
+            {
+                return this.NoContent();
```
Comment (Anny777) [inline Services/Utilities/Addressees/Addressees.API/Controllers/DistributionRegionController.cs:103]: return bad request

File: Services/Utilities/Addressees/Addressees.API/Controllers/DistributionRegionController.cs [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+            }
+
+            value.Pattern = value.Pattern.ToUpperInvariant();
+            var initializedValue = this.factory.MakeInstanceFrom(value);
+            var res = this.regionRepository.Create(initializedValue);
+            if (res != null)
+            {
+                return this.Created(new Uri($"/api/DistributionRegion?id={res.Id}", UriKind.Relative), new { res });
```
Comment (Anny777) [inline Services/Utilities/Addressees/Addressees.API/Controllers/DistributionRegionController.cs:111]: Return created at action

File: Services/Utilities/Addressees/Addressees.API/Controllers/DistributionRegionController.cs [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+        /// <response code="500">Unexpected error.</response>
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+        [HttpPatch]
+        public virtual ActionResult Patch([FromBody] Addressees.API.Models.DistributionRegion value)
+        {
+            if (value == null)
+            {
+                return this.NoContent();
```
Comment (Anny777) [inline Services/Utilities/Addressees/Addressees.API/Controllers/DistributionRegionController.cs:133]: return bad request

File: Services/Utilities/Addressees/Addressees.API/Migrations/20200911064302_Initial-Create.cs [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+            if (migrationBuilder == null)
+            {
+                throw new ArgumentNullException(nameof(migrationBuilder));
+            }
+
+            migrationBuilder.CreateTable(
+                name: "DistributionRegions",
+                columns: table => new
+                {
+                    Id = table.Column<string>(nullable: false),
```
Comment (Anny777) [inline Services/Utilities/Addressees/Addressees.API/Migrations/20200911064302_Initial-Create.cs:25]: Id must be Guid!

File: Services/Utilities/Addressees/Addressees.API/Migrations/20200911064302_Initial-Create.cs [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+                constraints: table =>
+                {
+                    table.PrimaryKey("PK_DistributionRegions", x => x.Id);
+                });
+
+            migrationBuilder.CreateTable(
+                name: "DistributionGroups",
+                columns: table => new
+                {
+                    Id = table.Column<string>(nullable: false),
```
Comment (Anny777) [inline Services/Utilities/Addressees/Addressees.API/Migrations/20200911064302_Initial-Create.cs:42]: Guid

File: Services/Utilities/Addressees/Addressees.API/Models/DistributionGroup.cs [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+        public virtual DistributionRegion DistributionRegion { get; set; }
+
+        /// <inheritdoc/>
+        public DistributionType DistributionType { get; set; }
+
+        /// <inheritdoc/>
+        public Division Division { get; set; }
+
+        /// <inheritdoc/>
+        [QuickSearch]
```
Comment (Anny777) [inline Services/Utilities/Addressees/Addressees.API/Models/DistributionGroup.cs:29]: Quick search delete

File: Services/Utilities/Addressees/Addressees.API/Models/DistributionGroup.cs [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+{
+    using Addressees.API.Abstraction.Enum;
+    using Addressees.API.Abstraction.Models;
+    using Tenders.Core.Models.Attributes;
+    using Tenders.Core.Models.Base;
+
+    /// <summary>
+    /// Model for distributing emails.
+    /// </summary>
+    public class DistributionGroup : ModelBase, IDistributionGroup
```
Comment (Anny777) [inline Services/Utilities/Addressees/Addressees.API/Models/DistributionGroup.cs:15]: Model base delete

File: Services/Utilities/Addressees/Addressees.API/Models/DistributionGroupViewModel.cs [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+
+    /// <summary>
+    /// View model for distribution group.
+    /// </summary>
+    public class DistributionGroupViewModel
+    {
+        /// <summary>
+        /// Gets or sets the identifier.
+        /// </summary>
+        public string Id { get; set; }
```
Comment (Anny777) [inline Services/Utilities/Addressees/Addressees.API/Models/DistributionGroupViewModel.cs:17]: Guid

File: Services/Utilities/Addressees/Addressees.API/Models/DistributionGroupViewModel.cs [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+
+        /// <summary>
+        /// Gets or sets the name of the distribution region.
+        /// </summary>
+        public string DistributionRegion { get; set; }
+
+        /// <summary>
+        /// Gets or sets the name of the distribution region id.
+        /// </summary>
+        public string DistributionRegionId { get; set; }
```
Comment (Anny777) [inline Services/Utilities/Addressees/Addressees.API/Models/DistributionGroupViewModel.cs:27]: Guid

File: Services/Utilities/Addressees/Addressees.API/Models/DistributionRegion.cs [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+{
+    using Addressees.API.Abstraction.Models;
+    using Tenders.Core.Models.Attributes;
+    using Tenders.Core.Models.Base;
+
+    /// <inheritdoc/>
+    public class DistributionRegion : ModelBase, IDistributionRegion
+    {
+        /// <inheritdoc/>
+        [QuickSearch]
```
Comment (Anny777) [inline Services/Utilities/Addressees/Addressees.API/Models/DistributionRegion.cs:15]: Delete all quick search attributes

File: Services/Utilities/Addressees/Addressees.API/Models/FilterOptions.cs [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+{
+    using System.Collections.Generic;
+    using System.Linq;
+    using Newtonsoft.Json;
+
+    /// <summary>
+    /// Filter options.
+    /// </summary>
+    public class FilterOptions
+    {
```
Comment (Anny777) [inline Services/Utilities/Addressees/Addressees.API/Models/FilterOptions.cs:15]: Delete this object,  search will be via devextreme.

File: Services/Utilities/Addressees/Addressees.API/ServiceRegistrationPoint.cs [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+    using Tenders.Core.Abstraction;
+    using Tenders.Core.Models.Base;
+
+    /// <inheritdoc/>
+    public class ServiceRegistrationPoint : ServiceRegistrationPointBase
+    {
+        /// <inheritdoc/>
+        public override void Register(IContainerConfig config, IServiceCollection sc)
+        {
+            RegisterModelBaseFactory<IDistributionGroup, DistributionGroup>(sc);
```
Comment (Anny777) [inline Services/Utilities/Addressees/Addressees.API/ServiceRegistrationPoint.cs:19]: Delete all used model base factory.

File: Services/Utilities/Addressees/Addressees.API/Abstraction/IRepository.cs [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+{
+    using System;
+    using System.Collections.Generic;
+    using Tenders.Core.Abstraction.Models.Base;
+
+    /// <summary>
+    /// Repository with base functionality.
+    /// </summary>
+    /// <typeparam name="TEntity">Model Type.</typeparam>
+    public interface IRepository<out TEntity>
```
Comment (Anny777) [inline Services/Utilities/Addressees/Addressees.API/Abstraction/IRepository.cs:15]: For what abstract repository? 
Is it possible to describe these methods in needing repositories?

File: Services/Utilities/Addressees/Addressees.API/Abstraction/IDistributionGroupRepository.cs [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+        /// Updates the specified item.
+        /// </summary>
+        /// <param name="item">The item.</param>
+        /// <returns>Operation result.</returns>
+        bool Update(DistributionGroupViewModel item);
+
+        /// <summary>Deletes the specified identifier.</summary>
+        /// <param name="id">The identifier.</param>
+        /// <returns>Operation result.</returns>
+        bool Delete(string id);
```
Comment (Anny777) [inline Services/Utilities/Addressees/Addressees.API/Abstraction/IDistributionGroupRepository.cs:57]: Id must be guid

File: Services/Utilities/Addressees/Addressees.API/Services/DistributionGroupRepository.cs [hunk 21]
```diff
@@ -0,0 +1,294 @@
...
+
+        /// <inheritdoc/>
+        public DistributionGroup Create(DistributionGroupViewModel item)
+        {
+            if (item == null)
+            {
+                return null;
+            }
+
+            var newInstance = this.distributionGroupFactory.MakeInstance();
```
Comment (Anny777) [inline Services/Utilities/Addressees/Addressees.API/Services/DistributionGroupRepository.cs:58]: May be via new? And used factory deleted

File: Services/Utilities/Addressees/Addressees.API/Services/DistributionGroupRepository.cs [hunk 22]
```diff
@@ -0,0 +1,294 @@
...
+            newInstance.DistributionRegion = region;
+            newInstance.DistributionType = item.DistributionType;
+            newInstance.Email = item.Emails;
+            newInstance.IsEnabled = item.IsEnabled;
+            newInstance.Division = item.Division;
+
+            newInstance.IsActive = true;
+            newInstance.CreatedDate = DateTime.Now;
+
+            this.quickSearchService.GenerateQuickSearchString(newInstance);
```
Comment (Anny777) [inline Services/Utilities/Addressees/Addressees.API/Services/DistributionGroupRepository.cs:75]: delete generate quick search

**Review by Anny777 — CHANGES_REQUESTED** [review]

File: Services/Utilities/Addressees/Addressees.Client/DistributionGroupClient.cs [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+        /// Gets receivers by division.
+        /// </summary>
+        /// <param name="division">Division.</param>
+        /// <returns>Receivers emails.</returns>
+        public async Task<IEnumerable<string>> GetAddresseesByDivision(Division division)
+        {
+            var uri = new Uri(this.getReceiversByDivision + division);
+            var client = this.httpClientFactory.CreateClient(uri);
+
+            var response = await Policy.Handle<HttpRequestException>()
```
Comment (Anny777) [inline Services/Utilities/Addressees/Addressees.Client/DistributionGroupClient.cs:106]: Added using on response

File: Services/Utilities/Addressees/Addressees.Client/DistributionRegionClient.cs [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+        /// Gets distribution region by code.
+        /// </summary>
+        /// <param name="regionCode">Region code.</param>
+        /// <returns>Supplier determination type.</returns>
+        public async Task<string> GetDistributionRegion(int regionCode)
+        {
+            var uri = new Uri(this.distribution + regionCode);
+            var client = this.httpClientFactory.CreateClient(uri);
+
+            var response = await Policy.Handle<HttpRequestException>()
```
Comment (Anny777) [inline Services/Utilities/Addressees/Addressees.Client/DistributionRegionClient.cs:53]: Added using on response

**Review by Anny777 — CHANGES_REQUESTED** [review]

File: UI/src/app/modules/settings/components/distribution/services/distribution.service.ts [hunk 25]
```diff
@@ -41,7 +43,7 @@ export class DistributionService implements IServiceBase<DistributionGroupRow, D
   }
 
   getEnumKeyValuePairs(): Observable<KeyValuePair<number, string>[]> {
-    return this.http.get<KeyValuePair<number, string>[]>(this.controllerUrl + "GetEnum");
+    return this.http.get<KeyValuePair<number, string>[]>(this.controllerUrl + 'GetDistributionEnum');
```
Comment (tatianaIljina) [inline UI/src/app/modules/settings/components/distribution/services/distribution.service.ts:46]: GetEnum

**Review by Anny777 — APPROVED** [review]

**Pull request merged**